sound/soc/sof/ipc4-loader.c

```diff
@@ -256,26 +254,111 @@ static int sof_ipc4_load_library_by_uuid(struct snd_sof_dev *sdev,
 	if (ret)
 		goto release;
 
-	ret = xa_insert(&ipc4_data->fw_lib_xa, lib_id, fw_lib, GFP_KERNEL);
+	ret = xa_insert(&ipc4_data->fw_lib_xa, *lib_id, fw_lib, GFP_KERNEL);
 	if (unlikely(ret))
 		goto release;
 
-	kfree(fw_filename);
-
+	(*lib_id)++;
 	return 0;
 
 release:
 	release_firmware(fw_lib->sof_fw.fw);
 	/* Allocated within sof_ipc4_fw_parse_ext_man() */
 	devm_kfree(sdev->dev, fw_lib->modules);
-free_filename:
-	kfree(fw_filename);
 free_fw_lib:
 	devm_kfree(sdev->dev, fw_lib);
 
 	return ret;
 }
 
+/**
+ * sof_ipc4_complete_split_release - loads the library parts of a split firmware
+ * @sdev: SOF device
+ *
+ * With IPC4 the firmware can be a single binary or a split release.
+ * - single binary: only the basefw
+ * - split release: basefw and two libraries (openmodules, debug)
+ *
+ * With split firmware release it is also allowed that for example only the
+ * debug library is present (the openmodules content is built in the basefw).
+ *
+ * To handle the permutations try to load the openmodules then the debug
+ * libraries as optional ones after the basefw boot.
+ *
+ * The libraries for the split release are stored alongside the basefw on the
+ * filesystem.
+ */
+int sof_ipc4_complete_split_release(struct snd_sof_dev *sdev)
+{
+	static const char * const lib_bundle[] = { "openmodules", "debug" };
+	const char *fw_filename = sdev->pdata->fw_filename;
+	const char *lib_filename, *p;
+	unsigned long lib_id = 1;
+	char *lib_name_base;
+	int i;
+
+	p = strstr(fw_filename, ".ri");
+	if (!p || strlen(p) != 3) {
+		dev_info(sdev->dev,
+			 "%s: Firmware name '%s' is missing .ri extension\n",
+			 __func__, fw_filename);
+		return 0;
+	}
+
+	lib_name_base = kzalloc(strlen(fw_filename) - 2, GFP_KERNEL);
+	if (!lib_name_base)
+		return -ENOMEM;
+
+	strscpy(lib_name_base, fw_filename, sizeof(lib_name_base));
```

Comment:
can you check the size please?

Reply:
there's only one size - the last argument of `strscpy()`. It's a size of a pointer, is that what is needed here?

Reply:
valid, it did worked on most platform with sof-123.ri pattern ;)